lib/df-detection.c

```diff
@@ -0,0 +1,275 @@
+#include "linux/df-detection.h"
+#include <asm/syscall.h>
+#include <linux/debugfs.h>
+#include <linux/fs.h>
+#include <linux/init.h>
+#include <linux/kernel.h>
+#include <linux/random.h>
+#include <linux/sched.h>
+#include <linux/slab.h>
+#include <linux/stacktrace.h>
+#include <linux/sys.h>
+#include <linux/types.h>
+
+void add_address(const void *addr, size_t len, unsigned long caller,
+		 void *kernel_addr)
+{
+	if (!in_task() || current->addresses == NULL || current->pairs == NULL ||
+	    addr > TASK_SIZE)
+		return;
+	char buf[64], buf_caller[64];
+	int size;
+	size = scnprintf(buf, sizeof(buf), "%ps", caller);
+	if (len > MAX_LEN || addr == 0 || strnstr(buf, "copy_from_user_nmi", size))
+		return;
+	if (current->num_read >= current->sz && current->sz < DF_MAX_RECORDS) {
+		struct df_address_range *temp =
+		    (struct df_address_range *)krealloc(
+			current->addresses,
+			current->sz * 2 * sizeof(struct df_address_range),
+			GFP_KERNEL);
+		current->addresses = temp ? temp : current->addresses;
+		current->sz =
+		    current->addresses ? current->sz * 2 : current->sz;
+	}
+	if (current->num_read < current->sz) {
+		current->addresses[current->num_read].start_address = addr;
+		current->addresses[current->num_read].len = len;
+		current->addresses[current->num_read].caller = caller;
+		current->addresses[current->num_read].stack =
+		    df_save_stack(GFP_NOWAIT);
```

Comment:
Please use tabs for indentation, unless you need to fine-align the next line with a specific character on the previous line (e.g. for function parameters).